### ISSUE TITLE
Scala.js runtime race condition: "Cannot block for result to be set in JavaScript"

### ISSUE THREAD

**Issue by davesmith00000** [zio/zio#8125]
Hi!

I'm having a problem with the unsafe runtime being unstable in Scala.js, currently it seems to work for me about 2 times in 10. 

I'm using it a project I maintain called [Tyrian](https://github.com/PurpleKingdomGames/tyrian), an Elm-like web framework. Tyrian is built on Scala 3, Scala.js, and primarily the Cats Effect Typeclass hierarchy. However, since the API only use concrete types (i.e. no tagless final or anything like that), we also support ZIO and the experience is pretty good, we (seem to have) have a growing number of Tyrian+ZIO users who are enjoying themselves. :smile:

Everything was working fine until we refactored Tyrian's runtime and upgraded to CE 3.5. Tyrian is now significantly faster (about x3 faster) and the ZIO version is no longer behaving well. This is what the code looked like, nothing fancy - almost straight out the docs:

```scala
  val run: Task[Nothing] => Unit = runnable =>
    val runtime = Runtime.default

    Unsafe.unsafe { implicit unsafe =>
      runtime.unsafe.run(runnable).getOrThrowFiberFailure()
    }
```

(Note that the CE version isn't any cleverer, it's basically just `_.unsafeRunAndForget()`.)

I've had a look around for related issues, and the closest I can find is this one: https://github.com/zio/zio/issues/3278

The specific error I'm getting is as follows:

![image](https://github.com/zio/zio/assets/908709/1cddf119-9ec2-46bb-980e-7ec066482a7b)

> "Cannot block for result to be set in JavaScript"

I tracked that down to here: https://github.com/zio/zio/blob/series/2.x/core/js/src/main/scala/zio/internal/OneShot.scala#L46
Which is invoked here: https://github.com/zio/zio/blob/series/2.x/core/shared/src/main/scala/zio/Runtime.scala#L135

I'm guessing that the fiber hasn't had time to get going before `result.get()` is called and the whole thing bombs out.

I can force it to work reliably on my machine with this horrible hack: :see_no_evil: 

```scala
  private val runtime = Runtime.default

  val run: Task[Nothing] => Unit = runnable =>
    val runIt =
      () =>
        Unsafe.unsafe { implicit unsafe =>
          runtime.unsafe.run(runnable).getOrThrowFiberFailure()
        }

    window.setTimeout(runIt, 100)
```

Does anyone have any idea what might be going on here beyond my pure speculation? Better yet, any suggestions for a fix / workaround?

Many thanks in advance.

**Comment by adamgfraser** [comment]
@davesmith00000 When you call `run` you are attempting to synchronously execute your workflow. This necessarily requires blocking for the result of the workflow if it is not entirely synchronous which is not possible on Scala.js. For this reason the document for `run` specifically says `May fail on Scala.js if the effect cannot be entirely run synchronously.` So I would look at what the code is you are calling in `run` and why it is not entirely synchronous.

**Comment by davesmith00000** [comment]
Hi @adamgfraser, thank you for coming back to me.

Is there some version of `run` appropriate for async? I did try using `fork` but it seemed to make no difference.

What's interesting is that it only fails some of the time. When it works, it works perfectly. Feels like a timing issue more than a fundamental limitation. I did read the docs you mentioned but wasn't sure where that left me.

Can I also ask then what you mean by synchronous in this case? JS is single thread (as I'm certain you know), so do you mean that there cannot be any callbacks in the code's start up process? If that's the case, perhaps that might suggest a workaround where the program initialises synchronously, and kicks off the app as a side effect somehow? (I'm grasping a bit, I realise. 😅)

**Comment by adamgfraser** [comment]
I'm surprised that `fork` didn't work for you as it does not involve that code path. `runToFuture` should also work for you.

It failing some of the time is a reflection of the fundamental issue of trying to block for an asynchronous computation. If something in the computation is asynchronous then there is a race condition between the fiber completing and getting the result of the fiber, which would result in exactly the kind of non-determinism you are seeing.

I mean anything that causes the fiber being run to suspend instead of immediately returning its result. Javascript is single threaded but in functional effect systems like ZIO or Cats Effect code on Javascript can still be concurrent because fibers can cooperatively yield to each other. For example one fiber might wait on a promise which would cause that fiber to suspend and allow the runtime to use that same single fiber to execute another fiber, which might then complete that promise. So in the same way, if you are unsafely running some code and it suspends that would give other code the chance to execute, such as the code that is trying to get the result of the fiber here.

**Comment by davesmith00000** [comment]
Morning @adamgfraser - thanks for the explanation.

I did some more digging and realised I'd been mistaken about `fork`. :facepalm: 

When I switched to `fork` the page _visibly_ failed to load in the same way as when I used `run`. However on closer examination the page was running and didn't throw any exceptions, but in some cases certain elements were not always ready yet, which explains why the artificial delay created with `setTimeout` showed an improvement. This is because the app starts much more quickly now - a high quality problem! :smile: 

So long story short: You were quite right, switching to `fork` does resolve the problem:

```scala
  private val runtime = Runtime.default

  val run: Task[Nothing] => Unit = runnable =>
    Unsafe.unsafe { implicit unsafe =>
      runtime.unsafe.fork(runnable)
    }
```

Thanks again for your time, I'll close the issue.
